java/sage/Carny.java

```diff
@@ -926,25 +967,185 @@ private void applyWasteData(Airing wasteAir)
+    List<Airing> mappedAirings = buildMap.get(hash);
+    // We are using an array list to keep things in order so that we can do a binary search later.
+    // Most of the data with the exception of airings sorted by channel/time should already be
+    // sorted, and we will do a final sort on that one specifically.
+    if (mappedAirings == null)
+    {
+      // 20 is selected based on random data samples. Often we only have one to two airings per
+      // array which can make this a really inefficient selection, but more often, we are close to
+      // 16 airings. This is admittedly a super small optimization, but we know we are going to put
+      // at least one airing in here, so creating the array is always going to happen at least once
+      // and this might keep us from needing to create a new one to expand.
+      mappedAirings = new ArrayList<>(20);
+      buildMap.put(hash, mappedAirings);
+      mappedAirings.add(airing);
+    }
+    else
+    {
+      // If we might be adding a duplicate, it will be the last entry because we call this multiple
+      // times for the same airing and it might have redundant hashes.
+      if (mappedAirings.get(mappedAirings.size() - 1) != airing)
+      {
+        mappedAirings.add(airing);
+      }
+    }
+  }
+
+  // Re-build the airing map for wasted so we don't need to keep re-getting wasted from the wizard.
+  // Wasted will cache it's airing on the first get which also keeps the requests at a minimum.
+  // This also keeps the airing binary sort which is desirable.
+  private Map<Integer, Wasted[]> buildWastedMap(Airing[] airings, int bitShift)
+  {
+    Map<Integer, Airing[]> airsMap = buildMap(airings, bitShift);
+    Map<Integer, Wasted[]> wastedMap = new HashMap<>((int)(airsMap.size() / 0.75) + 1);
+
+    for (Map.Entry<Integer, Airing[]>  entry : airsMap.entrySet())
+    {
+      Airing airs[] = entry.getValue();
+      Wasted wasted[] = new Wasted[airs.length];
+      int w = 0;
+      for (int i = 0; i < airs.length; i++)
+      {
+        Wasted currWasted = wiz.getWastedForAiring(airs[i]);
+        if (currWasted != null)
+          wasted[w++] = currWasted;
+      }
+      // Somehow we lost one while re-building.
+      if (w != airs.length)
+        wasted = Arrays.copyOf(wasted, w);
+      wastedMap.put(entry.getKey(), wasted);
+    }
+    return wastedMap;
+  }
+
+  private Map<Integer, Airing[]> buildMap(Airing airings[], int bitShift)
+  {
+    Map<Integer, List<Airing>> buildMap = new HashMap<>();
+    for (int i = 0, airingsSize = airings.length; i < airingsSize; i++)
+    {
+      Airing airing = airings[i];
+      Show show = airing.getShow();
+      if (show == null)
+        continue;
+
+      int currentHash = show.title == null ? 0 : show.title.ignoreCaseHash;
+      addToMaps((currentHash >>> bitShift), airing, buildMap);
+
+      currentHash = airing.isFirstRun() ? Agent.FIRSTRUN_MASK : Agent.RERUN_MASK;
```

Comment:
After thinking about this more, I feel like its use case is too limited. It was originally there for consistency and because it could make a difference in some cases.  I'm going to remove it.